### PR TITLE
Support ANSI mode for `ToUnixTimestamp, UnixTimestamp, GetTimestamp, DateAddInterval`

### PR DESCRIPTION
**Pull request by res-life**
Contributes to #5119 

Supports ANSI mode for:
- ToUnixTimestamp
- UnixTimestamp
- GetTimestamp
- DateAddInterval

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

**Comment by res-life** [comment]
build

File: integration_tests/src/main/python/date_time_test.py [hunk 1]
```diff
@@ -13,14 +13,17 @@
...
-from asserts import assert_gpu_and_cpu_are_equal_collect, assert_gpu_fallback_collect
+from asserts import assert_gpu_and_cpu_are_equal_collect, assert_gpu_fallback_collect, assert_gpu_and_cpu_error
 from data_gen import *
 from datetime import date, datetime, timezone
 from marks import incompat, allow_non_gpu
 from pyspark.sql.types import *
 from spark_session import with_spark_session, is_before_spark_330
 import pyspark.sql.functions as f
 
+ANSI_ENABLE_CONF = {'spark.sql.ansi.enabled': 'true'}
```
Comment (jlowe) [inline integration_tests/src/main/python/date_time_test.py:24]: This is a duplicate of `ansi_enabled_conf` later in this file.

File: integration_tests/src/main/python/data_gen.py [hunk 2]
```diff
@@ -1030,3 +1030,10 @@ def append_unique_to_rows(rows):
     new_rows = append_unique_to_rows(collected)
     new_schema = StructType(existing_schema.fields + [StructField("uniq_int", IntegerType(), False)])
     return spark.createDataFrame(new_rows, new_schema)
+
+# add multiple confs
+def add_conf(*confs):
```
Comment (jlowe) [inline integration_tests/src/main/python/data_gen.py:1035]: Do we really need this?  We already have `copy_and_update` whichcan be used to accomplish the same thing and is used in many other places, e.g.:
```
combined_confs = copy_and_update(first_conf, second_conf, third_conf ...)
```

File: integration_tests/src/main/python/date_time_test.py [hunk 3]
```diff
@@ -13,14 +13,17 @@
...
+LEGACY_INTERVAL_CONF = {'spark.sql.legacy.interval.enabled': 'true'}
```
Comment (jlowe) [inline integration_tests/src/main/python/date_time_test.py:25]: Nit: This should be moved next to the other predefined confs in this file

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 4]
```diff
@@ -499,10 +504,17 @@ object GpuToTimestamp extends Arm {
...
-      dtype: DType): ColumnVector = {
+      dtype: DType,
+      failOnError: Boolean): ColumnVector = {
 
     // `tsVector` will be closed in replaceSpecialDates
     val tsVector = withResource(isTimestamp(lhs.getBase, sparkFormat, strfFormat)) { isTs =>
+      if(failOnError && !BoolUtils.isAllValidTrue(isTs)) {
+        // ANSI mode and has invalid value.
+        // CPU may throw `DateTimeParseException`, `DateTimeException` or `ParseException`
+        throw new IllegalArgumentException("Exception occurred when parsing timestamp in ANSI mode")
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:515]: Do we need a followup issue that we're not throwing a matching exception type here?

Comment (res-life) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:515]: I will add more tests when fixing this PR: https://github.com/NVIDIA/spark-rapids/issues/1556.
Do we need to distinguish these exceptions?  If so, let's do it in this PR https://github.com/NVIDIA/spark-rapids/issues/1556

**Comment by res-life** [comment]
build

File: integration_tests/src/main/python/date_time_test.py (hunk 1 above)
Comment (res-life) [reply integration_tests/src/main/python/date_time_test.py]: done

File: integration_tests/src/main/python/date_time_test.py (hunk 3 above)
Comment (res-life) [reply integration_tests/src/main/python/date_time_test.py]: Done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala (hunk 4 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:515]: > Do we need to distinguish these exceptions? 

I don't think it is high priority to fix, and I could see adding some more tests while still leaving the exception types unaddressed.  So if #1556 does not address the exception types, we should file a low priority issue to track after that.

**Review by jlowe — APPROVED** [review]

**Review by nartal1 — APPROVED** [review]

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala (hunk 4 above)
Comment (res-life) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:515]: Filed an issue to track: https://github.com/NVIDIA/spark-rapids/issues/5355

**Pull request merged**